src/number_theory/modular_forms/slash_invariant_forms.lean

```diff
@@ -0,0 +1,209 @@
+defining `modular_form` and `cusp_form`. We prove several instances for such spaces, in particular
+that they form a module.
+-/
+
+open complex upper_half_plane
+
+open_locale upper_half_plane
+
+noncomputable theory
+
+local prefix `↑ₘ`:1024 := @coe _ (matrix (fin 2) (fin 2) _) _
+
+local notation `GL(` n `, ` R `)`⁺ := matrix.GL_pos (fin n) R
+
+local notation `SL(` n `, ` R `)` := matrix.special_linear_group (fin n) R
+
+section slash_invariant_forms
+
+set_option old_structure_cmd true
+
+open modular_form
+
+variables (F : Type*) (Γ : out_param $ subgroup SL(2, ℤ)) (k : out_param ℤ)
+
+localized "notation f `∣[`:73 k:0, A `]` :72 := slash_action.map ℂ k A f" in slash_invariant_forms
+
+/--Functions `ℍ → ℂ` that are invariant under the `slash_action`. -/
+structure slash_invariant_form :=
+(to_fun : ℍ → ℂ)
+(slash_action_eq' : ∀ γ : Γ, to_fun ∣[k, γ] = to_fun)
+
+/--`slash_invariant_form_class F Γ k` asserts `F` is a type of bundled functions that are invariant
+under the `slash_action`. -/
+class slash_invariant_form_class extends fun_like F ℍ (λ _, ℂ) :=
+(slash_action_eq : ∀ (f : F) (γ : Γ),  (f : ℍ → ℂ) ∣[k, γ] = f)
+
+attribute [nolint dangerous_instance] slash_invariant_form_class.to_fun_like
+
+@[priority 100]
+instance slash_invariant_form_class.slash_invariant_form :
+   slash_invariant_form_class (slash_invariant_form Γ k) Γ k :=
+{ coe := (slash_invariant_form.to_fun),
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  slash_action_eq := slash_invariant_form.slash_action_eq' }
+
+variables {F Γ k}
+
+instance : has_coe_to_fun (slash_invariant_form Γ k) (λ _, ℍ → ℂ) := fun_like.has_coe_to_fun
+
+@[simp] lemma slash_invariant_form_to_fun_eq_coe {f : slash_invariant_form Γ k} :
+  f.to_fun = (f : ℍ → ℂ) := rfl
+
+@[ext] theorem slash_invaraint_form_ext {f g : slash_invariant_form Γ k} (h : ∀ x, f x = g x) :
+  f = g := fun_like.ext f g h
+
+/-- Copy of a `slash_invariant_form` with a new `to_fun` equal to the old one.
+Useful to fix definitional equalities. -/
+protected def slash_invariant_form.copy (f : slash_invariant_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  slash_invariant_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',}
+
+end slash_invariant_forms
+
+namespace slash_invariant_form
+
+open slash_invariant_form
+
+variables {F : Type*} {Γ : out_param $ subgroup SL(2, ℤ)} {k : out_param ℤ}
+
+@[priority 100, nolint dangerous_instance]
+instance slash_invariant_form_class.coe_to_fun [slash_invariant_form_class F Γ k] :
+  has_coe_to_fun F (λ _, ℍ → ℂ) := fun_like.has_coe_to_fun
+
+@[simp] lemma slash_action_eqn [slash_invariant_form_class F Γ k] (f : F) (γ : Γ) :
+   slash_action.map ℂ k γ ⇑f = ⇑f := slash_invariant_form_class.slash_action_eq f γ
+
+lemma slash_action_eqn' (k : ℤ) (Γ : subgroup SL(2, ℤ)) [slash_invariant_form_class F Γ k] (f : F) :
+∀ γ : Γ, ∀ z : ℍ, f (γ • z) = ((↑ₘγ 1 0 : ℝ) * z +(↑ₘγ 1 1 : ℝ))^k * f z :=
```

Comment:
```suggestion
lemma slash_action_eqn' (k : ℤ) (Γ : subgroup SL(2, ℤ)) [slash_invariant_form_class F Γ k] (f : F)
  (γ : Γ) (z : ℍ) :
  f (γ • z) = ((↑ₘγ 1 0 : ℝ) * z +(↑ₘγ 1 1 : ℝ)) ^ k * f z :=
```
Leading `∀` arguments should be made hypotheses (this gives a definitionally equal lemma statement and lets you drop one line fromm the proof).

Is there a reason you're coercing to `ℝ` rather than `ℂ`?  As it is, it seems like those matrix coefficients are getting coerced twice, so maybe it's better to short-circuit.